oshdb-api/src/main/java/org/heigit/bigspatialdata/oshdb/api/mapreducer/MapReducer.java

```diff
@@ -1992,4 +2048,82 @@ private void checkTimeout() {
+   * `osmType`) this can result in a large performance boost.</p>
+   *
+   * <p>Currently, the following two optimizations are performed (but more could be feasibly be
+   * added in the future:</p>
+   *
+   * <p><b>basic optimizations:</b> includes simple filter expressions witch can be directly
+   * transformed to an (and-chain) of OSHDB filters (like OSM Tags or Types</p>
+   *
+   * @param mapRed the mapReducer whis the given filter was already applied on.
+   * @param filter the filter to optimize.
+   * @param <O> the type of the mapReducer to optimize (can be anything).
+   * @return a mapReducer with the same semantics as the original one, after some optimizations
+   *         were applied.
+   */
+  private <O> MapReducer<O> optimizeFilters(MapReducer<O> mapRed, FilterExpression filter) {
+    // basic optimizations
+    mapRed = optimizeFilters0(mapRed, filter);
+    // more advanced optimizations that rely on analyzing the DNF of a filter expression
+    mapRed = optimizeFilters1(mapRed, filter);
+    return mapRed;
+  }
+
+  private <O> MapReducer<O> optimizeFilters0(MapReducer<O> mapRed, FilterExpression filter) {
+    // basic optimizations (“low hanging fruit”):
+    // single filters, and-combination of single filters, etc.
+    if (filter instanceof TagFilterEquals) {
+      OSMTag tag = this.getTagTranslator().getOSMTagOf(((TagFilterEquals) filter).getTag());
+      return mapRed.osmTag(tag);
+    }
+    if (filter instanceof TagFilterEqualsAny) {
+      OSMTagKey key = this.getTagTranslator().getOSMTagKeyOf(
+          ((TagFilterEqualsAny) filter).getTag());
+      return mapRed.osmTag(key);
+    }
+    if (filter instanceof TypeFilter) {
+      return mapRed.osmType(((TypeFilter) filter).getType());
+    }
+    if (filter instanceof AndOperator) {
+      return optimizeFilters0(optimizeFilters0(mapRed,
+          ((AndOperator) filter).getLeftOperand()),
+          ((AndOperator) filter).getRightOperand());
+    }
+    return mapRed;
+  }
+
+  private <O> MapReducer<O> optimizeFilters1(MapReducer<O> mapRed, FilterExpression filter) {
+    // more advanced optimizations that rely on analyzing the DNF of a filter expression
+    List<List<Filter>> filterNormalized = filter.normalize();
+    // collect all OSMTypes in all of the clauses
+    EnumSet<OSMType> allTypes = EnumSet.noneOf(OSMType.class);
+    for (List<Filter> andSubFilter : filterNormalized) {
+      EnumSet<OSMType> subTypes = EnumSet.of(OSMType.NODE, OSMType.WAY, OSMType.RELATION);
+      for (Filter subFilter : andSubFilter) {
+        if (subFilter instanceof TypeFilter) {
+          subTypes.retainAll(EnumSet.of(((TypeFilter) subFilter).getType()));
+        } else if (subFilter instanceof GeometryTypeFilter) {
+          subTypes.retainAll(((GeometryTypeFilter) subFilter).getOSMTypes());
+        }
+      }
+      allTypes.addAll(subTypes);
+    }
+    mapRed = mapRed.osmType(allTypes);
+    // (todo) intelligently group queried tags
```

Comment:
I think this TODO is open on purpose and not supposed to be solved as part of this PR?

Reply:
correct